cropduster/utils/paths.py

```diff
@@ -17,22 +18,9 @@
 def get_upload_foldername(file_name, upload_to='%Y/%m'):
     # Generate date based path to put uploaded file.
     file_field = FileField(upload_to=upload_to)
-    if not file_name:
-        file_name = 'no_name'
-    filename = file_field.generate_filename(None, file_name)
+    filename = file_field.generate_filename(None, file_name or 'no_name')
     filename = re.sub(r'[_\-]+', '_', filename)
-
     if six.PY2 and isinstance(filename, unicode):
         filename = filename.encode('utf-8')
-
     root_dir = os.path.splitext(filename)[0]
-    root_dir = dir_name = os.path.join(settings.MEDIA_ROOT, root_dir)
-    i = 1
-    while os.path.exists(dir_name):
-        if six.PY2:
-            dir_name = b'%s-%d' % (root_dir, i)
-        else:
-            dir_name = '%s-%d' % (root_dir, i)
-        i += 1
-    os.makedirs(dir_name)
-    return dir_name
+    return default_storage.get_available_name(root_dir)
```

Comment:
You hate to see it, but I think we'll have to ditch `get_available_name` here for something more like what we originally had. Since `S3BotoStorage.exists` returns `False` for directories, it'd have to be something like:

```python
parent_dir, _, basename = root_dir.rpartition('/')
i = 1
dir_name = basename
sub_dirs, _ = storage.listdir(parent_dir)
while dir_name in sub_dirs:
    dir_name = "%s-%d" % (basename, i)
    i += 1
return "%s/%s" % (parent_dir, dir_name)
```

This is vulnerable to a possible race condition, but getting rid of it would make the code more complicated than it's worth.

Reply:
Agreed & updated 👍